cli/tools/bundle/mod.rs

```diff
@@ -0,0 +1,868 @@
+        esbuild_rs::BuiltinLoader::Json,
+      ))),
+      deno_graph::Module::Wasm(_) => todo!(),
+      deno_graph::Module::Npm(module) => {
+        let package_folder = self
+          .npm_resolver
+          .as_managed()
+          .unwrap() // byonm won't create a Module::Npm
+          .resolve_pkg_folder_from_deno_module(module.nv_reference.nv())?;
+        let path = self
+          .node_resolver
+          .resolve_package_subpath_from_deno_module(
+            &package_folder,
+            module.nv_reference.sub_path(),
+            None,
+            ResolutionMode::Import,
+            NodeResolutionKind::Execution,
+          )?;
+        let url = path.clone().into_url()?;
+        let path = path.into_path()?;
+        let (media_type, _charset) =
+          deno_media_type::resolve_media_type_and_charset_from_content_type(
+            &url, None,
+          );
+        let contents = std::fs::read(path)?;
+        Ok(Some((contents, media_type_to_loader(media_type))))
+      }
+      deno_graph::Module::Node(_) => Ok(None),
+      deno_graph::Module::External(_) => Ok(None),
+    }
+  }
+
+  fn specifier_and_type_from_graph(
+    &self,
+    specifier: &ModuleSpecifier,
+  ) -> Result<Option<(ModuleSpecifier, esbuild_rs::BuiltinLoader)>, AnyError>
+  {
+    let graph = self.module_graph_container.graph();
+    let Some(module) = graph.get(specifier) else {
+      return Ok(None);
+    };
+    let (specifier, loader) = match module {
+      deno_graph::Module::Js(js_module) => (
+        js_module.specifier.clone(),
+        media_type_to_loader(js_module.media_type),
+      ),
+      deno_graph::Module::Json(json_module) => (
+        json_module.specifier.clone(),
+        esbuild_rs::BuiltinLoader::Json,
+      ),
+      deno_graph::Module::Wasm(_) => todo!(),
+      deno_graph::Module::Npm(module) => {
+        let package_folder = self
+          .npm_resolver
+          .as_managed()
+          .unwrap() // byonm won't create a Module::Npm
+          .resolve_pkg_folder_from_deno_module(module.nv_reference.nv())?;
+        let path = self
+          .node_resolver
+          .resolve_package_subpath_from_deno_module(
+            &package_folder,
+            module.nv_reference.sub_path(),
+            None,
+            ResolutionMode::Import,
+            NodeResolutionKind::Execution,
+          )?;
+        let url = path.clone().into_url()?;
+        let (media_type, _charset) =
+          deno_media_type::resolve_media_type_and_charset_from_content_type(
+            &url, None,
+          );
+        (url, media_type_to_loader(media_type))
+      }
+      deno_graph::Module::Node(_) => {
+        return Ok(None);
+      }
+      deno_graph::Module::External(_) => {
+        return Ok(None);
+      }
+    };
+    Ok(Some((specifier, loader)))
+  }
+}
+
+fn file_path_or_url(url: &Url) -> String {
+  if url.scheme() == "file" {
+    url.to_file_path().unwrap().to_string_lossy().to_string()
```

Comment:
Prefer using the path utilities in `deno_path_util::url_to_file_path` because it will give a nicer error on failure and it works in Wasm (which this code doesn't need to work in atm, but it might make it easier in the future).